### ISSUE TITLE
GNU C Statement Expressions trigger false -Wunreachable-code

### ISSUE THREAD

**Issue by nickdesaulniers** [llvm/llvm-project#63173]
```c
int foo (void) {
    int x = 42;
    ({
        int x = 43;
        return x;
    });
    return x;
}
```
```
<source>:3:5: warning: code will never be executed [-Wunreachable-code]
    3 |     ({
      |     ^~
    4 |         int x = 43;
      |         ~~~~~~~~~~~
    5 |         return x;
      |         ~~~~~~~~~
    6 |     });
      |     ~~
```

**Comment by shafik** [comment]
I can see this both ways, although if you attempt to assign a variable to the result in this case we obtain an error. OTOH this code gives an awful diagnostic:

```cpp
 int y = ({
        int x = 43;
        return x;
        x;
    });
```

godbolt: https://godbolt.org/z/P13zoW1TK

So I tend to agree something needs to be fixed. CC @AaronBallman

**Comment by AaronBallman** [comment]
Yeah, this is a false positive IMO. The statement expression will be evaluated, the final `return` (outside of the statement expression) is what's unreachable.
